### PR TITLE
Refactor requests to use react-query instead of redux

### LINKED ISSUE #425: "Verifying URL" stage is sending duplicate requests & stalling
**Description**:
When doing the URL verification, duplicate requests are being sent and I believe this is causing it to stall sometimes due to these duplicated requests.

**Acceptance criteria**:
- [ ] URL verifier should not send duplicate requests
- [ ] Verifying URL stage should not stall

### LINKED ISSUE #783: Bad request when sorting/filtering on study hierarchy
**Description**:
When sorting and filtering on ISIS study hierarchy and then reloading the page (or changing the view) an error 400 occurs. This happens because the filters that are being applied to the studies are also being applied to the investigations linked to the currently viewed instrument. For example, navigating to study hierarchy and sorting by name will work just fine with the following request being triggered:
```
https://scigateway-preprod.esc.rl.ac.uk/datagateway-api/studyinvestigations?order="study.name asc","id asc"&skip=0&limit=10&where={"investigation.id":{"in":[16,28,43,50,66,71,76,81,97,110,127,134,155,224,234,239]}}&include="investigation","study"
```
but then reloading the page will trigger the following query to the API:
```
https://scigateway-preprod.esc.rl.ac.uk/datagateway-api/investigations?order="study.name asc","id asc"&where={"investigationInstruments.instrument.id":{"eq":1}}&distinct="id"
```
which is obviously invalid because investigations do not have the study property. 

The fix should be relatively straightforward: keep the currently applied filters, just don't apply them to the investigations query.

**Acceptance criteria**:
- [ ] Sorting and filtering on study hierarchy, followed by a page refresh or a view change should work just fine!

### LINKED ISSUE #798: Incorrect end date dataKey value for DLS dataset card view
**Description**:
https://github.com/ral-facilities/datagateway/blob/56b415ebc19c0daf84b1226594c92ce25b29c0a7/packages/datagateway-dataview/src/views/card/dls/dlsDatasetsCardView.component.tsx#L213 shows that the dataKey for the end date filtering of DLS datasets is wrong. It is currently set to `END_DATE` when it should be `endDate`, to mimic the changes made in https://github.com/ral-facilities/datagateway/pull/663. This naturally results in a 400 bad request.

As the component is being overhauled in https://github.com/ral-facilities/datagateway/pull/701, it will be fixed by that branch.

**Acceptance criteria**:
- [ ] filtering by end date works correctly

### PR DESCRIPTION
**Pull request by louise-davies**
## Description

This PR refactors the data fetching & state management of fetched data to be performed by the react-query library rather than redux. This both fixes the duplicate queries issue, as well as offers performance benefits due to react-query implementing request caching for us. There are also a lot of performance optimisations in the Table and Card views by memoising components to ensure these expensive components rerender as little as possible. There are other additional fixes as detailed below:

This PR also closes an outstanding problem with filtering on the study hierarchy view, as detailed in #783.

This PR also closes an outstanding problem with filtering on the DLS dataset card view, as detailed in #798. The commit that addresses this is https://github.com/ral-facilities/datagateway/pull/701/commits/a5eb00fa81facdea65b793df0acad99780339ad0.

This PR also fixes a problem discovered in that programmatically setting filters via query params didn't update the filter textboxes visually.
## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] Check scigateway-preprod no longer displays the issue of the same queries being sent twice (easiest way is to sort a table and see how many requests are sent - 1 or 2).

## Agile board tracking
Closes #425
Closes #783
Closes #798

### CODE REVIEW

**Comment by louise-davies** [comment]
I spoke a bit too soon - the previous commit fixes the issue in datagateway-search but not in datagateway-dataview.

File: packages/datagateway-common/package.json [hunk 1]
```diff
@@ -75,7 +78,7 @@
     "prepare": "yarn tsc",
     "test:watch": "react-scripts test --env=jsdom --watch",
     "lint:js": "eslint --ext=tsx --ext=ts --ext=js --ext=jsx --fix ./src",
-    "tsc": "tsc -p tsconfig.lib.json",
+    "tsc": "tsc --watch -p tsconfig.lib.json",
```
Comment (sam-glendenning) [inline packages/datagateway-common/package.json:81]: I probably committed this accidentally. Adding a note here to remember to take it out when this is all over

File: packages/datagateway-common/src/api/datafiles.tsx [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    (params) => fetchDatafileDetails(apiUrl, params.queryKey[1]),
+    {
+      onError: (error) => {
+        handleICATError(error);
+      },
+    }
+  );
+};
+
+export const downloadDatafileQuery = (
```
Comment (sam-glendenning) [inline packages/datagateway-common/src/api/datafiles.tsx:350]: Another note for later: change this to `downloadDatafile` once the `downloadDatafile` function in `src/state/actions/datafiles.tsx` has been totally removed from use

**Comment by sam-glendenning** [comment]
E2E test `should be able to sort by multiple fields` on `card/isis/investigations.spec.ts` is currently failing both locally (using preprod data) and on this Github Actions CI run. This is due to a card erroneously appearing when it should be filtered out. Goel has had a look at this but it is being stubborn so leaving this note here for us to go back and have a look at it later.

**Comment by GoelBiju** [comment]
> E2E test `should be able to sort by multiple fields` on `card/isis/investigations.spec.ts` is currently failing both locally (using preprod data) and on this Github Actions CI run. This is due to a card erroneously appearing when it should be filtered out. Goel has had a look at this but it is being stubborn so leaving this note here for us to go back and have a look at it later.
> 
> EDIT: just seen that commit pass all of its e2e tests so ??? should still check things are okay locally

I had skipped the failing e2e test for the time being so all of them are passing at the moment.

**Comment by sam-glendenning** [comment]
> E2E test `should be able to sort by multiple fields` on `card/isis/investigations.spec.ts` is currently failing both locally (using preprod data) ~and on this Github Actions CI run~. This is due to a card erroneously appearing when it should be filtered out. Goel has had a look at this but it is being stubborn so leaving this note here for us to go back and have a look at it later.
> 
> EDIT: just seen that commit pass all of its e2e tests so ??? should still check things are okay locally

This is now fixed in https://github.com/ral-facilities/datagateway/pull/701/commits/8d6ac2469e48ce735e53db0db426e27e0c91d766. The test was setup with some strange filter values so changed these to allow for the expected behaviour to occur.

**Review by sam-glendenning — APPROVED** [review]

**Pull request merged**